### PR TITLE
add support for using GitHub app-based authentication

### PR DESCRIPTION
**Pull request by gdams**
# Checklist

- [ ] You added tests to cover the change
- [x] `mvn clean install` build and test completes
- [x] You changed or added to the documentation

### CODE REVIEW

File: adoptium-updater-parent/adoptium-datasources-parent/adoptium-http-client-datasource/src/main/kotlin/net/adoptium/api/v3/dataSources/GitHubAuth.kt [hunk 1]
```diff
@@ -1,22 +1,59 @@
...
+            val installationId = System.getenv("GITHUB_APP_INSTALLATION_ID")
+
+            return if (!appId.isNullOrEmpty() && !privateKey.isNullOrEmpty() && !installationId.isNullOrEmpty()) {
+                LOGGER.info("Using GitHub App for authentication")
+                val jwtToken = authenticateAsGitHubApp(appId, privateKey)
+                if (jwtToken != null) {
+                    val gitHubApp: GitHub = GitHubBuilder().withJwtToken(jwtToken).build()
+                    val appInstallation: GHAppInstallation = gitHubApp.getApp().getInstallationById(installationId.toLong())
+                    val appInstallationToken: GHAppInstallationToken = appInstallation.createToken().create()
+                    LOGGER.info("Using token: ${appInstallationToken.token}")
```
Comment (gdams) [inline adoptium-updater-parent/adoptium-datasources-parent/adoptium-http-client-datasource/src/main/kotlin/net/adoptium/api/v3/dataSources/GitHubAuth.kt:38]: reminder to remove this once it's ready to merge

File: adoptium-updater-parent/adoptium-datasources-parent/adoptium-github-datasource/src/main/kotlin/net/adoptium/api/v3/dataSources/github/GitHubAuth.kt [hunk 2]
```diff
@@ -1,22 +1,57 @@
...
-        fun readToken(): String? {
-            var token = System.getenv("GITHUB_TOKEN")
-            if (token.isNullOrEmpty()) {
-                token = System.getProperty("GITHUB_TOKEN")
+        fun getAuthenticationToken(): AuthInfo {
+            val appId = System.getenv("GITHUB_APP_ID")
+            val privateKey = System.getenv("GITHUB_APP_PRIVATE_KEY")
+            val installationId = System.getenv("GITHUB_APP_INSTALLATION_ID")
+
+            return if (!appId.isNullOrEmpty() && !privateKey.isNullOrEmpty() && !installationId.isNullOrEmpty()) {
```
Comment (johnoliver) [inline adoptium-updater-parent/adoptium-datasources-parent/adoptium-github-datasource/src/main/kotlin/net/adoptium/api/v3/dataSources/github/GitHubAuth.kt:33]: should use `appId?.isNullOrEmpty()`

Comment (gdams) [reply adoptium-updater-parent/adoptium-datasources-parent/adoptium-github-datasource/src/main/kotlin/net/adoptium/api/v3/dataSources/github/GitHubAuth.kt:33]: guessing I should change them all?

```java
return if (!appId?.isNullOrEmpty() && !privateKey?.isNullOrEmpty() && !installationId?.isNullOrEmpty()) {
```

File: adoptium-updater-parent/adoptium-datasources-parent/adoptium-github-datasource/src/main/kotlin/net/adoptium/api/v3/dataSources/github/GitHubAuth.kt [hunk 3]
```diff
@@ -1,22 +1,57 @@
...
+                LOGGER.info("Using GitHub App for authentication")
+                val token = authenticateAsGitHubApp(appId, privateKey, installationId)
+                if (token != null) {
+                    AuthInfo(token.token, AuthType.APP, token.expiresAt)
+                } else {
+                    LOGGER.error("Failed to authenticate as GitHub App")
+                    AuthInfo(null, AuthType.NONE, null)
```
Comment (johnoliver) [inline adoptium-updater-parent/adoptium-datasources-parent/adoptium-github-datasource/src/main/kotlin/net/adoptium/api/v3/dataSources/github/GitHubAuth.kt:40]: not sure if there is any point continuing executing at this point? Throw an error and remove NONE type?

Comment (gdams) [reply adoptium-updater-parent/adoptium-datasources-parent/adoptium-github-datasource/src/main/kotlin/net/adoptium/api/v3/dataSources/github/GitHubAuth.kt:40]: In the past we've always supported running the API without any tokens. I wanted to preserve that behavior?

**Review by johnoliver — APPROVED** [review]

**Pull request merged**